### ISSUE TITLE
Allow a custom HttpMessageHandler to support overriding the verification of Self-Signed SSL certificates.

### ISSUE THREAD

**Issue by ElasticCoder** [Azure/azure-cosmos-dotnet-v3#42]
With the V2 client now using the latest framework it is possible to override the HttpMessageHandler and override self-signed certificate checking.

```
	var handler = new HttpClientHandler();
	handler.ClientCertificateOptions = ClientCertificateOption.Manual;
	handler.ServerCertificateCustomValidationCallback += DangerousAcceptAnyServerCertificateValidator;

	_client = new DocumentClient(new Uri(_documentDbOptions.Endpoint), _documentDbOptions.Key, handler);
```
However with the V3 client, there is no ability to do this. 
Can we look for a suitable way to do this?

Self-Signed certification verification overrides are required when running on Linux clients that wish to connect to the emulator. Despite all of the good work that has been done with the emulator such as allowing you to specify the alternative subject names for the self-signed certificate and export it, there are still limitations. I've found that the Linux implementation of .NET core that uses cURL/OpenSSL will still error with a self-signed certificate, even if you install it into the ca-certificates store. (Windows will honor the certificate if you put it into the Trusted Certificate Authorities store).

It would be preferable to use the V3 SDK rather than the V2.

**Comment by ElasticCoder** [comment]
I'm having a look through the source code to see how this might be done. 

Initially, it looks difficult. The example above uses the HttpClientHandler which derives from the abstract HttpMessageHandler. The Cosmos DB client framework uses a different derivative, HttpRequestMessageHandler which derives from a DelegatingHandler and the abstract HttpMessageHandler. 

Unfortunately this chain does not have access to the same ClientCertificateOptions and ServerCertificateCustomValidationCallback properties.

**Comment by ElasticCoder** [comment]
I've got a working solution. I'll start to prepare for a pull request.

**Comment by ElasticCoder** [comment]
@j82w would you like to discuss this?

**Comment by j82w** [comment]
I'm worried that there are other scenarios that will be blocked by not having the same access in v2. This will approach will not scale nicely if multiple other options need to be exposed.

@kirankumarkolli do you know if this is the only option that will be needed or do we need to find a better way to expose the different options?

**Comment by ElasticCoder** [comment]
@j82w Thanks for replying. Can you clarify what do you mean by "options"?

Are you concerned that the DisableSslVerification property is too specific to HTTP Gateway connections and will cause property bloat? Should it be encapsulated within a connection abstraction?

**Comment by j82w** [comment]
I am not to worried about the property bloat since Java and Python SDK have it.

I'm more interested in is if there are other scenarios that are being blocked by not having the ability to override the [HttpClientHandler](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclienthandler?view=netframework-4.7.2).  It exposes a lot of properties which the [HttpMessageHandler](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpmessagehandler?view=netstandard-2.0) does not have. If users need access to all of those properties that are no longer available then we should find a better way to expose it.

**Comment by Schueh** [comment]
@ElasticCoder Our local development environment is behind a proxy. That's one scenario where we need to override the HttpClientHandler to set a proxy. We already do that for Azure Storage and Search client SDKs. But I couldn't find a way in the new Azure Cosmos SDK.

**Comment by ElasticCoder** [comment]
> 
> 
> @ElasticCoder Our local development environment is behind a proxy. That's one scenario where we need to override the HttpClientHandler to set a proxy. We already do that for Azure Storage and Search client SDKs. But I couldn't find a way in the new Azure Cosmos SDK.

Thank you, that is a good concrete example. I'll have a look at the code impact.

**Comment by ElasticCoder** [comment]
I may have found a way to inject an HttpClientHandler into the DocumentClient which should satisfy the requirements that @Schueh and @j82w have raised. I will have a look at this next week.

**Comment by ElasticCoder** [comment]
I have addressed this requirement now with #74 which allows the consumer to specify a custom HttpMessageHandler.

**Comment by zhengwia** [comment]
Hi any plan on this pr going to release? Same situation working behind a corporate firewall. Thanks.

**Comment by kirankumarkolli** [comment]
.net standard 2.0 now seems has supports for custom [SSL handling](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclienthandler.servercertificatecustomvalidationcallback?view=netstandard-2.0#System_Net_Http_HttpClientHandler_ServerCertificateCustomValidationCallback).  

@ausfeldt started investigating and will update with details.

**Comment by kirankumarkolli** [comment]
@ElasticCoder we have a PR with the required changes. 
If possible can you please validate it?

**Comment by zhengwia** [comment]
Saw the pr merged to master! thanks guys looking forward to the next build!

**Comment by galvesribeiro** [comment]
Folks, how is that PR fixing the SSL bypass issue? I don't see how to set the `HttpClientHandler` or other way to ignore the SSL cert. I'm on OSX with the same issue on the emulator...

Can someone shed a light on this? Thanks!

**Comment by j82w** [comment]
Just in case anyone finds this issue it was fixed in [3.12.0](https://github.com/Azure/azure-cosmos-dotnet-v3/releases/tag/3.12.0).

Here are examples on how to configure it:
```
CosmosClientOptions cosmosClientOptions = new CosmosClientOptions()
{
    HttpClientFactory = () =>
    {
        HttpMessageHandler httpMessageHandler = new HttpClientHandler()
        {
            ServerCertificateCustomValidationCallback = (req, cert, chain, errors) => true
        };
        return new HttpClient(httpMessageHandler);
    },
    ConnectionMode = ConnectionMode.Gateway
};


CosmosClient client = new CosmosClient("https://localhost:8081", 
    "C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw==", 
    cosmosClientOptions);
```

Or if you are using a NET Standard 2.1 project:

```
CosmosClientOptions cosmosClientOptions = new CosmosClientOptions()
{
    HttpClientFactory = () =>
    {
        HttpMessageHandler httpMessageHandler = new HttpClientHandler()
        {
            ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator 
        };
        return new HttpClient(httpMessageHandler);
    },
    ConnectionMode = ConnectionMode.Gateway
};


CosmosClient client = new CosmosClient("https://localhost:8081", 
    "C2y6yDj", 
    cosmosClientOptions);
```
